Mathlib/SetTheory/Ordinal/Clubs.lean

```diff
@@ -0,0 +1,243 @@
+/-
+Copyright (c) 2024 Nir Paz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nir Paz
+-/
+import Mathlib.SetTheory.Cardinal.Cofinality
+
+
+/-!
+# Club and stationary sets
+This file sets up the basic theory of clubs (closed and unbounded sets) and stationary sets.
+
+## Main definitions
+* `IsClosed`: A set of ordinals is closed in `o` if all of its elements are
+  less than `o` and it contains all of its accumulation points below `o`.
+* `IsClub`: A set of ordinals is a club in `o` if it is closed in `o` and unbounded in `o`.
+
+## Main results
+* `isClub_sInter`: The intersection of fewer than `o.cof` clubs in `o` is a club in `o`.
+-/
+
+noncomputable section
+
+open Classical Cardinal Set Order
+
+universe u
+
+namespace Ordinal
+
+/-- A set of ordinals is unbounded in an ordinal if it
+  is a subset of it and it is an unbounded set in it. -/
+def IsUnbounded (S : Set Ordinal) (o : Ordinal) : Prop :=
+  S ⊆ (Iio o) ∧ (∀ p < o, ∃ s ∈ S, p < s)
+
+/-- A positive ordinal is an accumulation point of a set of ordinals if there
+  are elements in the set arbitrarily close to the ordinal from below. -/
+def IsAcc (o : Ordinal) (S : Set Ordinal) : Prop :=
+  0 < o ∧ (∀ p < o, ∃ s ∈ S, s < o ∧ p < s)
+
+/-- A set of ordinals is closed in an ordinals if it is a subset of it
+  and it contains all of its accumulation points below the ordinal. -/
+def IsClosed (S : Set Ordinal) (o : Ordinal) : Prop :=
+  S ⊆ (Iio o) ∧ (∀ p < o, IsAcc p S → p ∈ S)
+
+/-- A set of ordinals is a club in an ordinal if it is closed and unbounded in it. -/
+def IsClub (S : Set Ordinal) (o : Ordinal) : Prop :=
+  IsClosed S o ∧ IsUnbounded S o
+
+theorem isAcc_of_subset {o : Ordinal} {S T : Set Ordinal} (h : S ⊆ T) (ho : o.IsAcc S) :
+    o.IsAcc T := ⟨ho.1, fun p plto ↦ (ho.2 p plto).casesOn fun s hs ↦ ⟨s, h hs.1, hs.2⟩⟩
+
+theorem isLimit_of_isUnbounded {o : Ordinal} {U} (h : IsUnbounded U o)
+    (ho : o ≠ 0) : IsLimit o := by
+  refine' isLimit_of_not_succ_of_ne_zero (fun ⟨x, hx⟩ ↦ _) ho
+  rcases h.2 x (hx ▸ lt_succ x) with ⟨p, hp⟩
+  exact (hx ▸ succ_le_iff.mpr hp.2).not_lt (h.1 hp.1)
+
+theorem inter_suffix_of_isUnbounded {o : Ordinal} {U} (hU : IsUnbounded U o) {p : Ordinal}
+    (hp : p < o) : (U ∩ (Ioo p o)).Nonempty :=
+  (hU.2 p hp).casesOn (fun x ⟨xmemU, pltx⟩ ↦ ⟨x, xmemU, ⟨pltx, hU.1 xmemU⟩⟩)
+
+
+section ClubIntersection
+
+variable {o : Ordinal.{u}} {S : Set (Set Ordinal)}
+
+theorem isClosed_sInter_of_isClosed (hS : S.Nonempty) (h : ∀ C ∈ S, IsClosed C o) :
+    IsClosed (⋂₀ S) o :=
+  ⟨fun s hs ↦ by
+    rcases hS with ⟨C, CmemS⟩
+    exact (h C CmemS).1 ((sInter_subset_of_mem CmemS) hs),
+  fun p plto pAcc ↦ by
+    rw [mem_sInter]
+    intro C CmemS
+    exact (h C CmemS).2 p plto (isAcc_of_subset (sInter_subset_of_mem CmemS) pAcc)⟩
```

Comment:
```suggestion
    IsClosed (⋂₀ S) o where
  mp s hs := by
    rcases hS with ⟨C, CmemS⟩
    exact (h C CmemS).1 ((sInter_subset_of_mem CmemS) hs)
  mpr p plto pAcc := by
    rw [mem_sInter]
    intro C CmemS
    exact (h C CmemS).2 p plto (isAcc_of_subset (sInter_subset_of_mem CmemS) pAcc)
```
(personally, I think that's much nicer when the `⟨_, _⟩` would otherwise span several lines)